### PR TITLE
Refactor HangingNodes::setup_constraints()

### PR DESCRIPTION
**Pull request by peterrum**
With this PR, the algorithm is split up three steps:

https://github.com/dealii/dealii/blob/4d3c4090d6eaa12ebba0e8da6ff94113a052702b/include/deal.II/matrix_free/hanging_nodes_internal.h#L730-L755

(In follow up PRs, I would like to call the new functions at separate places to minimize redundant work).

The determination of the refinement configuration now consists of basic bit shifts of the form:

https://github.com/dealii/dealii/blob/4d3c4090d6eaa12ebba0e8da6ff94113a052702b/include/deal.II/matrix_free/hanging_nodes_internal.h#L439

https://github.com/dealii/dealii/blob/4d3c4090d6eaa12ebba0e8da6ff94113a052702b/include/deal.II/matrix_free/hanging_nodes_internal.h#L466

depends on ~#12979, #12980, #1298~

### CODE REVIEW

File: include/deal.II/matrix_free/hanging_nodes_internal.h [hunk 1]
```diff
@@ -384,386 +521,249 @@ namespace internal
...
+        compute_component_mask(cell->get_dof_handler().get_fe_collection());
+
+      if ([](const auto &outer) {
+            for (const auto &inner : outer)
+              for (const auto i : inner)
+                if (i)
+                  return true;
+            return false;
+          }(component_masks) == false)
+        return false;
```
Comment (kronbichler) [inline include/deal.II/matrix_free/hanging_nodes_internal.h:743]: I think this is hard to follow, why not simply
```suggestion
      return std::none_of(component_mask.begin(), component_mask.end(),
          [] (const auto &a) { return *std::max_element(a.begin(), a.end());});
```
Or write out the loops to find an entry with true (at the expense of an additional variable as we can't return from within the loop.

**Comment by peterrum** [comment]
@kronbichler I have made the changes. I also have added a test that the algorithm works for (2D) mixed meshes (with a few additional changes).

File: include/deal.II/matrix_free/hanging_nodes_internal.h [hunk 2]
```diff
@@ -345,24 +380,142 @@ namespace internal
...
+
+          // ignore if at boundary
+          if (cell->at_boundary(face_no))
+            continue;
+
+          const auto &neighbor = cell->neighbor(face_no);
+
+          // ignore neighbors that are artificial or have the same level or
+          // have children
+          if (cell->neighbor(face_no)->has_children() ||
```
Comment (kronbichler) [inline include/deal.II/matrix_free/hanging_nodes_internal.h:449]:
```suggestion
          if (neighbor->has_children() ||
```

**Review by kronbichler — APPROVED** [review]

File: include/deal.II/matrix_free/hanging_nodes_internal.h (hunk 2 above)
Comment (peterrum) [reply include/deal.II/matrix_free/hanging_nodes_internal.h]: Done!
